man/h2man.1
.TH H2MAN 1 "1 Jan 2000" "NODElib" "NODElib Programmer's Manual"
.SH NAME
h2man \- generate man pages from header files.
.SH SYNOPSIS
.B h2man
[\-title name] [\-section num] [\-source name]
.br
\ \ \ \ \ \ [\-manual name] files ... [options]

But usually just:

.B h2man
file.h
.SH DESCRIPTION
The
.B h2man
program is a
.B perl\c
(1)
script that generates a
.B man\c
(1) page from the supplied C header files.  The header files must obey
a few simple rules which should not be too difficult for most
programmers to follow. The ideal setting for
.B h2man
is in documenting the programmer's interface to a collection of C
routines.  The output is written to a file named with the same
base-name as the input file, but with a suffix given by the section
number (or a default of 3) as is standard for man pages.
.SS Options
The following options only effect the very first line of the generated
man page, which selects the title and sectioning parameters for the
man page.  Each option corresponds exactly with one option to the
".TH" command used in the
.B nroff -man
format.  See
.B man\c
(7) for more details.
.TP
.I -title name
The title of the man page (e.g., 
.I H2MAN\c
).  The default is the base-name of the processed file translated into
capital letters.
.TP
.I -section num
The section  number the man page should be placed in and the file name
suffix to give the output file name (e.g.,
.I 1\c
).  The default is 3, since that is the section for program library
routines.
.TP
.I -source name
The source of the command (e.g.,
.I SCRLIB\c
).  This is usually the name of the package your header file will be
part of.  The default is the empty string.
.TP
.I -manual name
The title of the manual (e.g.,
.I SCRLIB Programmer's Manual\c
).  The default is the empty string.
.SH FILE FORMAT
The man page generated by
.B h2man
can contain any number of user specified sections, but it must always
contain at least NAME, SYNOPSIS, and DESCRIPTION which are
automatically generated from the contents of the header file. The
contents of the NAME section is simply taken from the header file's
first comment.  The SYNOPSIS section can contain any user specified
information, but it will always contain a list of all function
prototypes and preprocessor macros that take an argument.  The
DESCRIPTION section is subdivided into as many as four subsections:
type declarations, function prototypes, macros, and miscellany.  Each
subsection contains documentation for only those portions of the
header file that have been explicitly documented, i.e. if you do not
comment a function prototype it will not appear in the DESCRIPTION
section.  The miscellany subsection contains documentation from
comments which begin with
.B h2man:include
and is intended primarily for global variable documentation.

.B H2man
groups all comments in a header file into three different types
which are primarily grouped according to the comment's location
relative to C code:
.TP
.B A file header
Each header file must have one header comment which is the very first
comment in the file.
.TP
.B Interface documentation
Any comment which immediately precedes a function prototype, a typedef
declaration, or a preprocessor macro which takes at least one
argument. Also, any comment which begins with
.B h2man:include.\c
\.
.TP
.B  Ignored comments
Any other type of comment, or one which is explicitly ignored because
of a directive.
.LP
.SS File Header
The smallest possible file header looks like:
.LP
.nf
/* NAME
 *   foo.h
 */
.fi
.LP
But a more descriptive one is:
.LP
.nf
/* Copyright (c) 1995 by Gary W. Flake.  All rights reserved.
 * Please see a the release notes for the full conditions.
 *
 * NAME
 *   foo.h - an example for h2man.
 * SYNOPSIS
 *   Does very little but illustrates some of the capabilities
 *   of the program.  In particular, the text in this section
 *   will be placed in front of the function prototypes in the
 *   synopsis section.
 * DESCRIPTION
 *   Likewise, this text will be the first in the description
 *   section.
 * ANOTHER SECTION
 *   This will be yet another section which starts after the
 *   description section.
 */
.fi
.LP
Notice that the comment is flush-left with each line starting with a
"\ *\ ".  Currently, all file headers must have the same structure.
However, this may change if enough people complain.
The copyright notice is optional because
.B h2man
throws away all text of the file header up until the first line that
contains a word all in capital letters.  The amount of indentation is
not too crucial, but all non-section names must be indented at least
one space relative to the section names, which must be on their own
line and in all capital letters.
.SS Interface Documentation
For any function prototype, typedef type declaration, or preprocessor
macro which takes and argument you can document the code with a
comment that starts in the first column and precedes the related C
code.  Interface comments should not have a flush-left "\ *\ " border,
as does the file header.

Function prototypes should also be flush-left as well. Moreover,
prototypes should contain argument names in addition to types.
Otherwise,
.B h2man
will do a poor job of converting the prototype into a suitable form
for viewing.

For documenting anything besides a typedef, prototype, or macro you
can use the
.B h2man:include
directive which looks like:
.nf

/* h2man:include - Any documentation that you would want
   to include in the Miscellaneous subsection of the
   description section.  This is a good place to document
   globals. */
.fi
.SS  Ignored Comments
As a rules of thumb, any comment which precedes anything but a typedef,
prototype, or macro will be ignored.  Thus, a simple place to put a
comment invisible to
.B h2man
is before a global variable or another comment.  You can also
explicitly tell
.B h2man
to ignore comments as well as whole sections of code.  Here are the
different ways of doing this.
.TP
.B Skipping a single line:
.nf
/* h2man:ignore */ int ignores_just_this_line_line;

int ignores_just_this_line_line; /* h2man:ignore */
.fi
.TP
.B Skipping one syntactical expression
The next two directives cause
.B h2man
to skip only the prototype and the typedef expression because the
directive is on its own line and with no additional text inside the
comment.
.nf

/* h2man:ignore */
int foo(int bar);

/* h2man:ignore */
typedef foo_s {
  int a;
  char b;
} FOO;
.fi
.TP
.B Skipping whole blocks:
.nf
/* h2man:skipbeg */

  /* Any number of comments */
  /* Any number of C expressions */

/* h2man:skipend */

.fi
Whenever
.B h2man
strips the header file based on comment directives, the header file
is processed as if the directives and the removed text never existed.
Thus, you should watch out for cases such as:
.nf

/* Some unrelated comment */

/* h2man:ignore */
int foo(void);
int bar(void);

.fi
In the example above the unrelated comment will be coupled to the
prototype for bar(), which is clearly not the intention.
.SS Escape Codes and Embedded Commands.
All text that is passed on to the man page is filtered for escape
sequences.  The backslash, "\\", can be used to quote special
characters or it can be used as "\\n" to add an embedded newline into
the text.  All carriage returns remain intact in the text unless
escaped at the end of line with a backslash.

By embedding
.B nroff\c
(1) commands into the comments you can add a little personality to the
documentation. For example:
.nf

/* This is a \\n.B bold\\n word, and this is an
   .I underlined\\n word. */
.fi
.LP
would be appear as:
.LP

This is a 
.B bold
word.  This is an 
.I underlined
word.
.LP
after being processed by
.B h2man
and
.B nroff.
.SS Itemized lists
One could embed the commands necessary to compose an itemized list,
but this would result in a C comment that would be very unreadable.
As an alternative,
.B h2man
grabs anything that looks like an itemized list and processes it in a
sane way.  A line looks like an itemized list if it begins with one of
"*o+-" or "0-9" (with a possible trailing ".").  If the item extends
to more than one line then the carriage returns must be escaped. For
example:
.nf

/* This is the start of my list:
     * This is the first item.
     * This is the second item which extends to multiple \\
       lines and just keeps going and going.
     * This is the third item. */
.fi
.LP
would turn into:
.LP
.PD 0
This is the start of my list:
.TP
.B *
This is the first item.
.TP
.B *
This is the second item which extends to multiple line and just keeps
going and going.
.TP
.B *
This is the third item.
.PD 1
.SH BUGS
Undoubtedly some exist.  It would be nice to be able to determine if
comments have or do not have a flush-left border of some sort, instead
of blindly following the author's preference.  Some of the pattern
matching code is probably too simplistic as well.
.SH AUTHOR
Gary W. Flake (\c
.B gary.flake@usa.net\c
).
.SH SEE ALSO
.B nroff\c
(1),
.B troff\c
(1),
.B groff\c
(1),
.B man\c
(1)
.B man\c
(7),
.B perl\c
(1).

etc/NEWS
Nodelib NEWS:

1.12.0:  The union of several DATASETs can be treated as one by using
         the DSM_UNION type.

1.11.9:  Alias links supported but untested.

1.11.0:  Online kmeans added.  Thanks Chris.  API to nn_create_rbf()
         and nn_create_smlp() has been simplified (e.g., there are no
         more "random" and "normalized" version of these calls.)
         These options are now set to reasonable defaults controlled
         by library globals.  See kmeans.3 and the bottom of nn.3 for
         more details.

1.10.0:  Variable length delays for SERIES.  The program tseries.c 
         tests the implementation of this.

1.9.11:  Backprop!  Whoopeee!

1.9.10:  Supports SMLP architecture.  RBF, NRBF, and SMLP creation
         calls have variants that either use K-means clustering or
         random sampling of the input space.

1.9.4:   Weight decay for OPTIMIZERs.

1.9.0:   Normalized links can be used to normalize entire layers.  This
         new feature is used in the nn_create_nrbf() function which
         creates an RBF with normalized basis functions.

1.8.9:   NN_LINK for copy connection without backprop.

1.8.8:   DSM_FILE routines are now memory mapped under unix.

1.8.7:   Two new DATASET functions for returning data in your own
         space: dataset_x_copy() and dataset_y_copy().

1.8.0:   The DSM_FIFO DATASET_METHOD allows you to have a DATASET that
         only contains the most current data.  On creation, you decide
         how big the DSM_FIFO should be.  Each time you add a new
         pattern the oldest will overwritten.

1.7.0:   The DSM_SUBSET DATASET_METHOD allows you to make a temporary
         meta-dataset that consist of a subset of another DATASET.
         This is handy of you are doinf incremental-like learning on a
         huge set.

1.6.1:   A new SVD module is now available.  There is also a spinv()
         call which assumes a positive definite input matrix.

1.6.0:   Weightless pairwise product links.  CNLS network architecture
         is now supported with the new links.  An example CNLS network
         is given.  nn_create() and nn_link() now take a variable
         number of arguments, just like a printf().  You can use "%d"
         and extra arguments instead of doing your own sprintf() call
         for architectures that are not determined until runtime.

1.5.5:   Preliminary version nn_jacobian().  Fixed alignment bug in
         allocate_array().

1.5.4:   nn_solve() can now work on any linear link.  This makes
         incremental regression a snap.

1.5.3:   Hopfield network example solving a combinatorial optimization
         problem which uses the scalar link type from the previous
         release.

1.5.2:   A new scalar connection type is now supported.  Given two
         layers with the same number of nodes (named SOURCE and SINK,
         for example), we now have a connection that does
         SINK[i] = a[i] * SOURCE[i].  Note that SOURCE and SINK can be
         the same layer, so this allows for a type of recurrent
         memory.

1.5.0:   NODElib's SVD routine can now use LAPACK and BLAS.  This can
         result in an order of magnitude increase in speed for pca(),
         pinv(), and nn_solve() calls.  See the FAQ for more details.

1.4.0:   PCA now inplace for any DATASET type.

1.3.14:  Extension to h2man for \em{} and \bf{} instead of embedded
         nroff code.  This makes the header files more readable.

1.3.12:  pinv() is now based on a non NRC SVD routine.

1.3.10:  nn_solve() will skip patterns with targets set to NaN.

1.3.9:   FAQ list in HTML format.

1.3.8:   Testing and bug fixes for DSM_FILE code.

1.3.7:   DS_METHOD for large binary flat files, but untested.

1.3.6:   DS_METHOD for dynamic C matrices implemented for pointers to
         pointers to doubles.  The allocate_array() function now works
         with a single malloc call and keeps all data memory
         contiguous.  Test programs for modules will now be placed in
         the test subdirectory.  More to follow.

1.3.5:   The nn_solve() routine is now very memory efficient.  Storage
         is now independent of the DATASET size.

1.3.4:   Automagic documentation generation now handles elaborate
         function prototypes.

1.3.2:   Compiles clean under Windows.  Thanks WEB.

1.3.1:   Change in SCAN module now allows newlines as delimiters, per
         Rickard's request.

1.3.0:   Linear weights can now be solved for exactly with an SVD
         routine.  An RBFN can be created, have it's centers clustered,
         and its linear weights solved for exactly with a single
         function call.  

1.2.2:   BFGS and DFP quasi-Newton search.  GMK's hybrid line search in
         beta release. There are now five distinct optimization
         algorithms, each of which can use three different line search
         routines.

1.2.1:   OPTIMIZE type generalized for line searches.  Cubic line search
         is now supported.  Example is included.

1.2.0:   K-means clustering with example in cluster.c.  HTML directory
         now has an automagically generated index file.

1.1.2:   Most (all?) documentation is complete. Xor.c is now a verbosely
         documented example.

1.1.1:   Builds under Windows. C Matrices can now be used as DATASETs.

1.1.0:   PATTERN renamed to SERIES.  All of the function names for
         PATTERNs have changed as well.  DATASET is now the most generic
         data type for NNs.  SERIES is a special case of DATASET.

1.0.0:   First formal release.  Almost builds under Windows.  Now has
         hyperlinked HTML.


etc/h2man.pl

# Converts header files into man pages.  See this man page for
# details.  This is my first perl script ever, so please feel
# free to improve on it. -- Gary Flake.


######################################################################

# Returns the first location in a list of the occurance of a pattern.

sub first_loc {
  local($where);
  local($pat, *array) = ($_[0], $_[1]);
  undef $where;
  for ($[ .. $#array) {
    $where = $_, last if $array[$_] =~ /$pat/;
  }
  $where;
}

######################################################################

# Returns the first location in a list of the occurance of a pattern
# starting from some supplied starting point.

sub first_loc_from {
  local($where);
  local($pat, $beg, *array) = ($_[0], $_[1], $_[2]);
  undef $where;
  for ($beg .. $#array) {
    $where = $_, last if $array[$_] =~ /$pat/;
  }
  $where;
}

######################################################################

# Formats a chunk of text into nroff format.  Most of the work
# is in handling itemized lists.  The rest is just some simple
# reformatting.

sub do_text {
  local(*txt) = $_[0];
  local($item) = 0;
  local($verb) = 0;
  local($pref) = "";
  local($postf) = "";
  foreach(@txt) {
    s/^   //g;
    if(/\\begin{verbatim}/) {
      s/\\begin{verbatim}/.LP\n.nf\n/g;
      $verb = 1;
    }
    if(/\\end{verbatim}/) {
      s/\\end{verbatim}/.fi\n.LP/g;
      $verb = 0;
    }

    if(/\\bf *$/) {
      $pref = ".B \"";
      $postf = "\"";
      s/ *\\bf *[\n]//g;
    }
    elsif(/\\em *$/) {
      $pref = ".I \"";
      $postf = "\"";
      s/ *\\em *[\n]//g;
    }
    elsif(/\\rm *$/) {
      $pref = "";
      $postf = "";
      s/ *\\rm *[\n]//g;
    }

    s/^(..*)$/$pref\1$postf/g;
    s/^ *\n/.LP\n/g;

    s/ *\\begin{itemize} */.RS\n/g;
    s/ *\\end{itemize} *\n*/.RE\n/g;

    s/\\indent *(.*)/\n.RS\n\1\n.RE\n/g;
    s/^ *\\item */.br\n/g;
    s/\\it/\\em/g;
    s/\\newline/\n.br\n/g;
    s/\\section{([^}]*)}/.SS \1/g;

    # convert zzz\em{xxx}yyy to zzz\c\n.I xxx\c\nyyy
    s/([^ ])\\em\{([^\}]*)}([^ \n][^ \n]*)/\1\\\\c\\n.I \2\\\\c\\n\3/g;
    # convert \em{xxx}yyy to \n.I xxx\c\nyyy
    s/\\em\{([^\}]*)}([^ \n][^ \n]*)/\\n.I \1\\\\c\\n\2/g;
    # convert \em{xxx} to \n.I xxx\n
    s/\\em\{([^\}]*)}/\\n.I \1\\n/g;

    # convert zzz\bf{xxx}yyy to zzz\c\n.B xxx\c\nyyy
    s/([^ ])\\bf\{([^\}]*)}([^ \n][^ \n]*)/\1\\\\c\\n.B \2\\\\c\\n\3/g;
    # convert \bf{xxx}\n to \n.B xxx\n
    s/\\bf\{([^\}]*)}\\n/\\n.B \1\\n/g;
    # convert \bf{xxx}yyy to \n.B xxx\c\nyyy
    s/\\bf\{([^\}]*)}([^ \n][^ \n]*)/\\n.B \1\\\\c\\n\2/g;
    # convert \bf{xxx} to \n.B xxx\n
    s/\\bf\{([^\}]*)}/\\n.B \1\\n/g;

    # convert \url{anchor}{href}
    s/\\url\{([^\}]*)\}\{([^\}]*)\}/$1/g;

    # convert \qurl{href}
    s/\\qurl\{([^\}]*)\}/$1/g;

    s/\\n$//g;
    s/^\\n//g;
    s/\\n *\\n/\\n/g;
    s/(.)\\n *\\n/\1\\n/g;

    if($verb == 0) {
      s/^\s\s*([^\s].*)/\1/g;	# Remove leading spaces.
    }
    s/\\n\s*/\n/g;		# Insert requested newlines
    s/(.*[^\\])\\\n/\1/g;	# Remove "\\\n".
    if($verb == 0) {
      s/^\s\s*(.*\n)/\1/g;	# Remove leading white-space.
    }
    s/\\(.)/\1/g;		# Replace any remaining "\X" with "X".
    s/ - / \\- /g;		# Add "\" in front of "-".
    s/^ *\\-/\\-/g;

    # Fix broken period.
    s/\n\.  *(.*)\n/\n.RI . \n\1\n/g;
    s/\n\.  *\n/\n.RI . \n/g;
    s/\n\.$/\n.RI . \n/g;

    s/\n\n\n*/\n/g;

    print OUTFILE $_;
  }
}

######################################################################

# Formats a whole section and removes it from the header comment.

sub do_section {
  local($sname) = $_[0];

  $i = &first_loc("^$sname", *header);
  if(!defined($i)) { return 0; }
  @tmp = @header[$i+1 .. $#header];
  $j = &first_loc("^[A-Za-z]+", *tmp);
  if(!defined($j)) { $j = $#tmp + 1; }
  @tmp = splice(@header, $i, ($j + 1));
  print OUTFILE ".SH $tmp[0]";
  @tmp = @tmp[1 .. $#tmp];
  foreach(@tmp) { s/^/ /g; }
  $item = 0;
  &do_text(*tmp);
  return 1;
}

######################################################################

sub convert_sub_proto {
    local($_) = $_[0];
    if(/^\.BI.*/) {
	s/^\.BI //g;
	s/([^\\]) /\1/g;
	s/\\(.)/\1/g;
    }
    return($_);
}


######################################################################

# Converts a prototype into a form such that nroff will bold types,
# and underline arguments.  Yes, this is disgusting, but it
# apparently works, and that's good enough for me right now.

sub convert_proto {
  local($proto) = $_[0];
  local($munge) = 0;
  local(@plines);
  local($i);
  $_ = $proto;
  s/[ \t]+/ /g;
  # strip out function pointer arg lists.
#  s/(\(\*\w*\))\([^\)]*\)/\1\(\)/g;
  if(length($_) > 56) {
    $munge = $_;
#    $munge =~ s/^([^\(]*)(\s|\*)\w+\(.*/\1/g;
    $munge =~ s/^([^\(]*)\(.*/\1/g;
    $munge = length($munge);
  }

  s/[ \t]+/\\ /g;
  s/(\w+),/ \1 ,\\ /g;
  s/(\w+)\)/ \1 )/g;
  s/\( void \)/(void)/g;
  s/\\ \\ /\\ /g;

  if($munge != 0) {
    $munge = '\ ' x $munge;
    s/^([^\(]+\([^\(]+)\s*,\\\s*\\*\s*/\1 ,\n.br\n.BI $munge/g;
    s/ ,\\ ([^\)]+\);)/ ,\n.br\n.BI $munge\1/g;
    s/\)\s*,\\\s*/),\n.br\n.BI $munge/g;

    if(/.*\(\* \w+ \)\(.*/) {

      @plines = split("\n", $_);
      $i = 0;
      foreach(@plines) {
        $psub = &convert_sub_proto($_);
	if(length($psub) > 62 && /.*\(\* \w+ \)\(.*/) {
	    $munge2 = $psub;
#	    $munge2 =~ s/(.*\)\()[^\(]*/\1/g;
#	    $munge2 =~ s/([^\(]+\(\*).*/\1/g;
            $munge2 =~ s/([^\(]+\(\*\w+\)\().*/\1/g;
	    $munge2 = length($munge2);
	    $munge2 = '\ ' x $munge2;
	    $_ = $plines[$i];
	    while(/ ,\\ ([^\),]+)/) {
		s/ ,\\ ([^\),]+)/ ,\n.br\n.BI $munge2\1/g;
	    }
	    $plines[$i] = $_;
        }
        $i++;
      }
      $_ = join("\n", @plines);
      $_ = "$_\n";
    }
  }

  $_;
}

######################################################################

# Kills empty lines and selected lines.  Types of kills:
#
# Kills just the one line.
# anything ... /* h2man:ignore */ anything-else on this line...
#
# Kills just the comment.
# /* h2man:ignore anything-else inside the comment... */
#
# Kills the next function prototype;
# /* h2man:ignore */
# foo();
# 
# Kills the next typedef.
# /* h2man:ignore */
# typedef ...
#
# Kills a whole block.
# /* h2man:skipbeg */
# any number of lines of anything
# /* h2man:skipend */

sub clean_rest {
  local($beg, $skip);
  local($i) = 0;
  local($len) = $#rest;
  for($i = 0; $i < $len; $i++) { # join line with /*\*/ seperating them.
    $_ = $rest[$i];
    if(/\/\*\\\*\/\s*$/) {
      $rest[$i] =~ s/\/\*\\\*\/\s*//g;
      $rest[$i] = $rest[$i] . $rest[$i+1];
      splice(@rest, $i+1, 1);
      $_ = $rest[$i];
      if(!/\/\*\\\*\/\s*$/) {
	$i++;
      }
      else {
	$i--;
      }
    }
  }
  for($i = 0; $i < $len; $i++) { # kill borderlike comments.
    $_ = $rest[$i];
    if(/^\s*\/\*\W*\*\/\s*$/) {
      splice(@rest, $i, 1);
    }
  }
  undef $beg;
  while(1) { # clean up h2man:ignore directives.
    $beg = &first_loc('.*\/\*\s*h2man:ignore.*', *rest);
    if(!defined($beg)) { last; }
    if($rest[$beg] =~ /^\/\*\s*h2man:ignore\s*\*\/.*\S+.*/) {
      splice(@rest, $beg, 1);
    }
    elsif($rest[$beg] =~ /.*\S+.*\/\*\s*h2man:ignore\s*\*\/.*/) {
      splice(@rest, $beg, 1);
    }
    elsif($rest[$beg] =~ /^\/\*\s*h2man:ignore.*\w+.*/) {
      $end = &first_loc_from('^.*\*\/.*', $beg, *rest);
      splice(@rest, $beg, $end - $beg + 1);
    }
    elsif($rest[$beg + 1] =~ /^\s*typedef.*/) {
      if($rest[$beg + 1] =~ /.*;.*/) {
	$end = $beg + 1;
      }
      else {
	$end = &first_loc_from('^\s*\}\s*\w+\s*;.*', $beg + 1, *rest);
      }
      if(!defined($end)) {
	splice(@rest, $beg, 1);
      }
      else {
	splice(@rest, $beg, $end - $beg + 1);
      }
    }
    elsif($rest[$beg + 1] =~ /^(\w(\w|\s|\*)+)\([^\)]*\)\s*;/) {
      splice(@rest, $beg, 2);
    }
    else {
      splice(@rest, $beg, 2);
    }
    undef $beg; undef $end;
  }
  @tmp = ();
  $skip = 0;
  $inside = 0;
  foreach(@rest) { # kill blank lines.
    if($skip && /^\s*\/\*\s*h2man:skipend.*/) {
      $skip = 0;
    }
    elsif(/^\s*\/\*\s*h2man:skipbeg.*/) {
      $skip = 1;
    }
    elsif(/.*\/\*.*\*\/\s*/) {
      @tmp = (@tmp, $_);
    }
    elsif(/.*\/\*.*/) {
      $inside = 1;
      if($skip == 0) {
	@tmp = (@tmp, $_);
      }
    }
    elsif(/.*\*\/.*/) {
      $inside = 0;
      if($skip == 0) {
	@tmp = (@tmp, $_);
      }
    }
    elsif((!/^\s*$/) && $skip == 0) {
      @tmp = (@tmp, $_);
    }
    elsif(/^\s*$/ && $skip == 0 && $inside == 1) {
      @tmp = (@tmp, $_);
    }
  }
  @rest = @tmp;
}

######################################################################

# Processes all typedefs

sub do_typedefs {
  local($cur, $beg, $end, $i, @com, @td, $first);
  undef $beg;
  
  $cur = 0;
  $first = 0;
  while(1) {
    undef $beg;
    $beg = &first_loc_from('^typedef.*', $cur, *rest);
    if(!defined($beg)) { return; }
    if($rest[$beg] =~ /.*;.*/) {
      $end = $beg;
    }
    else {
      $end = &first_loc_from('^}\s*\w+\s*;\s*', $beg + 1, *rest);
    }

    if($rest[$beg-1] =~ /.*\*\/\s*$/) {
      @td = @rest[$beg .. $end];
      $i = $beg - 1;
      while(!($rest[$i] =~ /^\/\*.*/)) {
	$i--;
      }
      @com = @rest[$i .. $beg - 1];
      $com[0] =~ s/.*\/\*(.*\n)/  \1/g;
      $com[$#com] =~ s/(.*)\*\/.*\n/\1\n/g;
      if($first == 0) {
        print OUTFILE ".SS Type Declarations\n.LP\n";
	print OUTFILE "The following types are defined in the header ";
	print OUTFILE "file \\c\n.IR $inname .\n";
	$first = 1;
      }
      print OUTFILE ".LP\n.nf\n";
      foreach (@td) {
        chop($_);
        print OUTFILE ".BI \"$_\"\n";
      }
      print OUTFILE ".fi\n.LP\n";
      &do_text(*com);
    }
    $cur = $end + 1;
  }
}

######################################################################

# Processes all prototypes for the synopsis section.

sub do_protos {
  local($cur, $beg, $end, $i, @com, $proto, $first, $okay);
  undef $beg;
  
  $cur = 0;
  $first = 0;
  while(1) {
    undef $beg;
    $okay = 0;
    while($okay != 1) {
      $beg = &first_loc_from('^\w+(\s|\*)+\w+(\w|\s|\*)*\(.*\)\s*;',
			     $cur, *rest);
      if(!defined($beg)) { return; }
      if(@rest[$beg] =~ /^\s*typedef/) {
	$cur = $beg + 1;
      }
      else {
        $okay = 1;
      }
    }
    $end = $beg;
    if($rest[$beg-1] =~ /.*\*\/\s*$/) {
      $proto = $rest[$beg];
      $i = $beg - 1;
      while(!($rest[$i] =~ /^\/\*.*/)) {
	$i--;
      }
      @com = @rest[$i .. $beg - 1];
      $com[0] =~ s/.*\/\*(.*\n)/  \1/g;
      $com[$#com] =~ s/(.*)\*\/.*\n/\1\n/g;
      if($first == 0) {
        print OUTFILE ".SS Function Definitions\n.LP\n";
	print OUTFILE "The following function prototypes are given ";
	print OUTFILE "in the header file \\c\n.IR $inname .\n";
	$first = 1;
      }
      print OUTFILE ".LP\n.nf\n";
      $proto = &convert_proto($proto);
      print OUTFILE ".BI $proto.br\n.fi\n";
      &do_text(*com);
    }
    $cur = $end + 1;
  }
}

######################################################################

# Processes all macros.

sub do_macros {
  local($cur, $beg, $end, $i, @com, $macro, $first);
  undef $beg;
  
  $cur = 0;
  $first = 0;
  while(1) {
    undef $beg;
    $beg = &first_loc_from('^#define \w+\([^\)]*\)', $cur, *rest);
    if(!defined($beg)) { return; }
    $end = $beg;
    if($rest[$beg-1] =~ /.*\*\/\s*$/) {
      $macro = $rest[$beg];
      $i = $beg - 1;
      while(!($rest[$i] =~ /^\/\*.*/)) {
	$i--;
      }
      @com = @rest[$i .. $beg - 1];
      $com[0] =~ s/.*\/\*(.*\n)/  \1/g;
      $com[$#com] =~ s/(.*)\*\/.*\n/\1\n/g;
      if($first == 0) {
        print OUTFILE ".SS Macro Definitions\n.LP\n";
	print OUTFILE "The following macros are defined ";
	print OUTFILE "in the header file \\c\n.IR $inname .\n";
	$first = 1;
      }
      print OUTFILE ".LP\n.nf\n";
      $macro = &convert_macro($macro);
      print OUTFILE ".BI $macro.br\n.fi\n";
      &do_text(*com);
    }
    $cur = $end + 1;
  }
}

######################################################################

# Takes comments of the form: /* h2man:include ... */ and puts
# the entire contents into a misc subsection under description.
# use this for globals.

sub do_extras {
  local($cur, $beg, $end, $i, @com, $first);
  undef $beg;
  
  $cur = 0;
  $first = 0;
  while(1) {
    undef $beg;
    $beg = &first_loc_from('^\/\*\s*h2man:include.*', $cur, *rest);
    if(!defined($beg)) { return; }
    $end = &first_loc_from('.*\*\/', $beg, *rest);
    @com = @rest[$beg .. $end];
    $com[0] =~ s/.*\/\*\s*h2man:include\s+(.*\n)/\1/g;
    $com[$#com] =~ s/(.*)\*\/.*\n/\1\n/g;
    if($first == 0) {
      print OUTFILE ".SS Miscellaneous Items\n.LP\n";
      $first = 1;
    }
    &do_text(*com);
    $cur = $end + 1;
  }
}

######################################################################

# converts a macro into a pretty printing form.

sub convert_macro {
  local($macro) = $_[0];
  $_ = $macro;
  s/^(#define) (\w+\([^\)]*\)).*/\1\\ \2 .../g;
  s/\(\s*/\( /g;
  s/\s*\)\s*/ \)\\ /g;
  s/\s*,\s*/ , \\ /g;
  $_;
}
    
######################################################################

# Does one whole file.

sub do_file {
  if($filter = 0) {
    open(INFILE, $inname) || die "h2man: can't open \"$inname\".\n";
    open(OUTFILE, ">$outname") || die "h2man: can't open \"$outname\".\n";  
  }
  else {
    open(INFILE, "-");
    open(OUTFILE, ">-");
  }
  
  @lines = <INFILE>;

  # Get the header comment
  $beg = &first_loc('^\s*\/\*', *lines);
  $end = &first_loc('^\s*\*\/', *lines);
  @header = @lines[$beg .. $end];
  if($header[0] =~ /^\s*\/\* [A-Z][A-Z]+\s*/) {
    $header[0] =~ s/^\s*\// /g;
  }
  else {
    splice(@header, 0, &first_loc('^ \* ([A-Z][A-Z]+\s)+', *header));
    pop(@header);
  }
  foreach(@header) {
    s/ \* (.*)/\1/g;
    s/^ \*$//g;
  }

  # Get everything but the header
  @rest = @lines[$end + 1 .. $#lines];
  &clean_rest();

  # Write out the starting info
  $date = `date '+%d %h %Y'`;
  chop($date);
  print OUTFILE ".TH $title $section \"$date\" \"$source\" \"$manual\"\n";

  if(! &do_section('NAME')) {
    print OUTFILE ".SH NAME\n$inname.\n";
  }

  if(! &do_section('SYNOPSIS')) {
    print OUTFILE ".SH SYNOPSIS\n";
  }
  @protos = grep(/^\w+(\s|\*)+\w+(\w|\s|\*)*\(.*\)\s*;/, @rest);
  @protos = grep(!/^ *typedef/, @protos);
  print OUTFILE ".LP\n.B #include <$inname>\n.LP\n";
  foreach(@protos) {
    if(/^\s*private/i) {
      next;
    }
    $tmp = &convert_proto($_);
    print OUTFILE ".BI $tmp.br\n";
  }
  @macros = grep(/^#define \w+\([^\)]*\)/, @rest);
  foreach(@macros) {
    if(/private/i) {
      next;
    }
    $tmp = &convert_macro($_);
    print OUTFILE ".BI $tmp.br\n";
  }

  # Description = typedefs, protos, macros, and globals.
  if(! &do_section('DESCRIPTION')) {
    print OUTFILE ".SH DESCRIPTION\n";
  }

  &do_typedefs();
  &do_protos();
  &do_macros();
  &do_extras();

  @sects = grep(/^\w+.*/, @header);
  foreach(@sects) {
    &do_section($_);
  }

  close(INFILE);
  close(OUTFILE);
}

######################################################################

sub main_loop {
  local($iiii);			# perl bug.  if named $i it wont work.
  $inname = $outname = $title = $section = $source = $manual = "";
  
  $filter = 1;
  for($iiii = 0; $iiii < $#ARGV + 1; $iiii++) {
    $_ = $ARGV[$iiii];
    if(/^-title$/) {
      $title = $ARGV[$iiii + 1];
      $iiii++;
    }
    elsif(/^-section$/) {
      $section = $ARGV[$iiii + 1];
      $iiii++;
    }
    elsif(/^-source$/) {
      $source = $ARGV[$iiii + 1];
      $iiii++;
    }
    elsif(/^-manual$/) {
      $manual = $ARGV[$iiii + 1];
      $iiii++;
    }
    elsif(/^-name/) {
      $inname = $ARGV[$iiii + 1];
      $iiii++;
    }
    elsif(/.*\.h$/) {
      $filter = 1;
      if($section eq "") {
	$section = "3";
      }
      $outname = $inname = $_;
      $outname =~ s/(.*)\.h/\1.$section/g;
      if($title eq "") {
	$title = $inname;
	$title =~ s/(.*).h/\1/g;
	$title =~ tr/a-z/A-Z/;
      }
      &do_file();
      $inname = $outname = $title = $section = $source = $manual = "";
    }
    else {
      die "h2man: bad option \"$_\".\n";
    }
  }
  if($filter == 1) {
    &do_file();
  }	
}

######################################################################

&main_loop();

